misc/ls-extensions/modules/json-to-record-converter/src/main/java/io/ballerina/converters/JsonToRecordConverter.java

```diff
@@ -0,0 +1,400 @@
+import org.apache.commons.lang3.StringUtils;
+import org.ballerinalang.formatter.core.Formatter;
+import org.ballerinalang.formatter.core.FormatterException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static io.ballerina.compiler.syntax.tree.NodeFactory.createBuiltinSimpleNameReferenceNode;
+import static io.ballerina.converters.util.ConverterUtils.convertOpenAPITypeToBallerina;
+import static io.ballerina.converters.util.ConverterUtils.escapeIdentifier;
+import static io.ballerina.converters.util.ConverterUtils.extractReferenceType;
+
+/**
+ * Implements functionality to convert Json (object or schema) to a ballerina record.
+ *
+ * The conversion is done by using openAPI schema as an intermediary.
+ *
+ * @since 2.0.0
+ */
+public class JsonToRecordConverter {
+
+    private JsonToRecordConverter() {
+        // not called
+    }
+
+    /**
+     * This method takes in a json string and returns the Ballerina code block.
+     *
+     * @param jsonString json string for the schema
+     * @return {@link String} ballerina code block
+     * @throws IOException in case of Json parse error
+     * @throws ConverterException in case of invalid schema
+     * @throws FormatterException in case of invalid syntax
+     */
+    public static String convert(String jsonString) throws IOException,
+            ConverterException, FormatterException {
+        ObjectMapper objectMapper = new ObjectMapper();
+        OpenAPI model;
+        JsonNode inputJson = objectMapper.readTree(jsonString);
+        if (inputJson.has("$schema")) {
+            model = parseJSONSchema(jsonString);
+        } else {
+            Map<String, Object> schema = SchemaGenerator.generate(inputJson);
+            String schemaJson = objectMapper.writeValueAsString(schema);
+            model = parseJSONSchema(schemaJson);
+        }
+        ArrayList<TypeDefinitionNode> typeDefinitionNodeList = generateRecords(model);
+        NodeList<ImportDeclarationNode> imports = AbstractNodeFactory.createEmptyNodeList();
+        NodeList<ModuleMemberDeclarationNode> moduleMembers = AbstractNodeFactory.createNodeList(
+                typeDefinitionNodeList.toArray(new TypeDefinitionNode[0]));
+        Token eofToken = AbstractNodeFactory.createIdentifierToken("");
+        ModulePartNode modulePartNode = NodeFactory.createModulePartNode(imports, moduleMembers, eofToken);
+        TextDocument textDocument = TextDocuments.from("");
+        SyntaxTree syntaxTree = SyntaxTree.from(textDocument);
+        syntaxTree = syntaxTree.modifyWith(modulePartNode);
+        return Formatter.format(syntaxTree).toString();
+    }
+
+    /**
+     * Generates Ballerina Record nodes for given OpenAPI model.
+     *
+     * @param openApi OpenAPI model
+     * @return {@link ArrayList<TypeDefinitionNode>}  List of Record Nodes
```

Comment:
Can we use type arguments here? If not, we might have to add it as `{@link ArrayList}`.

Reply:
changed in 72c26d9